sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/implementation/directconnectivity/ConnectionStateListenerTest.java

```diff
@@ -0,0 +1,104 @@
+
+    private static int port = 8082;
+    private static String serverUriString = "rntbd://localhost:" + port;
+    private static final Logger logger = LoggerFactory.getLogger(ConnectionStateListenerTest.class);
+
+    @DataProvider(name = "connectionStateListenerConfigProvider")
+    public Object[][] connectionStateListenerConfigProvider() {
+        return new Object[][]{
+            // isTcpConnectionEndpointRediscoveryEnabled, serverResponseType, GlobalAddressResolver.updateAddresses() called times
+            {true, RequestResponseType.CHANNEL_FIN, 1},
+            {false, RequestResponseType.CHANNEL_FIN, 0},
+            {true, RequestResponseType.CHANNEL_RST, 0},
+            {false, RequestResponseType.CHANNEL_RST, 0},
+        };
+    }
+
+    @Test(groups = { "unit" }, dataProvider = "connectionStateListenerConfigProvider")
+    public void connectionStateListener_OnConnectionEvent(
+        boolean isTcpConnectionEndpointRediscoveryEnabled,
+        RequestResponseType responseType,
+        int times) throws ExecutionException, InterruptedException {
+
+        TcpServer server = TcpServerFactory.startNewRntbdServer(port);
+        // Inject fake response
+        server.injectServerResponse(responseType);
+
+        ConnectionPolicy connectionPolicy = new ConnectionPolicy(DirectConnectionConfig.getDefaultConfig());
+        connectionPolicy.setTcpConnectionEndpointRediscoveryEnabled(isTcpConnectionEndpointRediscoveryEnabled);
+
+        GlobalAddressResolver addressResolver = Mockito.mock(GlobalAddressResolver.class);
+
+        SslContext sslContext = SslContextUtils.CreateSslContext("client.jks", false);
+
+        Configs config = Mockito.mock(Configs.class);
+        Mockito.doReturn(sslContext).when(config).getSslContext();
+
+        RntbdTransportClient client = new RntbdTransportClient(
+            config,
+            connectionPolicy,
+            new UserAgentContainer(),
+            addressResolver);
+
+        RxDocumentServiceRequest req =
+            RxDocumentServiceRequest.create(mockDiagnosticsClientContext(), OperationType.Create, ResourceType.Document,
+                "dbs/fakedb/colls/fakeColls",
+                getDocumentDefinition(), new HashMap<>());
+        req.setPartitionKeyRangeIdentity(new PartitionKeyRangeIdentity("fakeCollectionId","fakePartitionKeyRangeId"));
+
+        Uri targetUri = new Uri(serverUriString);
+        try {
+            client.invokeStoreAsync(targetUri, req).block();
+        } catch (Exception e) {
+            logger.info("expected failed request with reason {}", e);
+        }
+        finally {
+            Mockito.verify(addressResolver, Mockito.times(times)).updateAddresses(Mockito.any(), Mockito.any());
+        }
+
+        TcpServerFactory.shutdownRntbdServer(server);
+    }
+
+    private Document getDocumentDefinition() {
+        String uuid = UUID.randomUUID().toString();
+        Document doc = new Document(String.format("{ "
```

Comment:
also can we replace this with ObjectNode? 

Some of the existing tests use Document but for new tests it is better to move to the new model.

Reply:
@moderakh  sorry Mo I merged the PR, I will change this in a following PR, thanks~